### DISCUSSION TITLE
Question: Swizzling Behavior in AdaptiveCpp vs. oneAPI (DPC++)

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1313] [Q&A]
Hello,

I recently encountered an interesting difference in behavior between AdaptiveCpp and oneAPI (DPC++) while working on a SYCL codebase. 

In my code, I have a `pos` variable of type `sycl::float2`, and I was performing a swizzling operation on a `velocity3d` vector of `sycl::float3` type as follows:

`pos += velocity3d.xy() * delta_t / cell_size;`

When compiling this code with AdaptiveCpp, I encountered an error related to the use of the * operator with the swizzled components, specifically stating that _data is a private member of 'hipsycl::sycl::vec<float, 2>'.

```
C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\sycl\libkernel\vec.hpp(630,3): error G898843A3: '_data' is a private member of 'hipsycl::sycl::vec<float, 2>'
    630 |   HIPSYCL_DEFINE_BINARY_VEC_OP_VEC_SCALAR(*, T)
        |   ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\sycl\libkernel\vec.hpp:619:14: note: expanded from macro 'HIPSYCL_DEFINE_BINARY_VEC_OP_VEC_SCALAR'
    619 |       result._data[i] = lhs._data[i] op rhs;                                   \
        |              ^
  D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:638:31: note: in instantiation of member function 'hipsycl::sycl::operator*' requested here
    638 |                                                         pos += velocity3d.xy() * delta_t / cell_size;
        |                                                                                ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\glue\generic\hiplike\hiplike_kernel_launcher.hpp:177:3: note: in instantiation of function template specialization 'hipsycl::glue::hiplike_dispatch::device_invocation_with_local_reducers<(lambda at C:\AdaptiveCpp\bin\..\include\AdaptiveCpp\hipSYCL\glue\cuda\..\generic\hiplike\hiplike_kernel_launcher.hpp:177:41)>' requested here
    177 |   device_invocation_with_local_reducers([&](auto& ... local_reducers){
        |   ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\glue\generic\hiplike\hiplike_kernel_launcher.hpp:215:7: note: in instantiation of function template specialization 'hipsycl::glue::hiplike_dispatch::device_invocation<(lambda at C:\AdaptiveCpp\bin\..\include\AdaptiveCpp\hipSYCL\glue\cuda\..\generic\hiplike\hiplike_kernel_launcher.hpp:209:3)>' requested here
    215 |       device_invocation([&] __host__ __device__(auto... reducers) {
        |       ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\glue\generic\hiplike\hiplike_kernel_launcher.hpp:732:35: note: in instantiation of function template specialization 'hipsycl::glue::hiplike_dispatch::parallel_for_kernel<__hipsycl_unnamed_kernel, (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36), 1>' requested here
    732 |                 hiplike_dispatch::parallel_for_kernel<kernel_name_t>, kernel_name_t,
        |                                   ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\glue\generic\hiplike\hiplike_kernel_launcher.hpp:868:27: note: in instantiation of function template specialization 'hipsycl::glue::hiplike_dispatch::invoke_reducible_kernel<(lambda at C:\AdaptiveCpp\bin\..\include\AdaptiveCpp\hipSYCL\glue\cuda\..\generic\hiplike\hiplike_kernel_launcher.hpp:709:41) &, hipsycl::glue::hiplike_dispatch::reduction_stage<1>>' requested here
    868 |         hiplike_dispatch::invoke_reducible_kernel(
        |                           ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\glue\kernel_launcher_factory.hpp:90:15: note: in instantiation of function template specialization 'hipsycl::glue::hiplike_kernel_launcher<hipsycl::rt::backend_id::cuda, hipsycl::rt::cuda_queue>::bind<hipsycl::glue::kernel_name_traits<__hipsycl_unnamed_kernel, (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36)>, hipsycl::rt::kernel_type::basic_parallel_for, 1, (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36)>' requested here
     90 |     launcher->bind<name_traits, Type>(offset, global_range, local_range,
        |               ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\sycl\handler.hpp:797:15: note: in instantiation of function template specialization 'hipsycl::glue::make_kernel_launchers<__hipsycl_unnamed_kernel, hipsycl::rt::kernel_type::basic_parallel_for, 1, (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36)>' requested here
    797 |         glue::make_kernel_launchers<KernelName, KernelType>(
        |               ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\sycl\handler.hpp:323:13: note: in instantiation of function template specialization 'hipsycl::sycl::handler::submit_kernel<__hipsycl_unnamed_kernel, hipsycl::rt::kernel_type::basic_parallel_for, (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36), 1>' requested here
    323 |       this->submit_kernel<KernelName, rt::kernel_type::basic_parallel_for>(
        |             ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\sycl\handler.hpp:329:13: note: in instantiation of function template specialization 'hipsycl::sycl::detail::separate_last_argument_and_apply<(lambda at C:\AdaptiveCpp\bin\..\include\AdaptiveCpp\sycl\..\hipSYCL\sycl\handler.hpp:322:20) &, const (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36) &>' requested here
    329 |     detail::separate_last_argument_and_apply(invoker, redu_kernel...);
        |             ^
  D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:7: note: in instantiation of function template specialization 'hipsycl::sycl::handler::parallel_for<__hipsycl_unnamed_kernel, (lambda at D:\source\repos\G2.cpp\G.SYCL\Debris\Debris.cpp:621:36)>' requested here
    621 |                                 h.parallel_for(particle_count, [=](auto i)
        |                                   ^
  C:\AdaptiveCpp\include\AdaptiveCpp\hipSYCL\sycl\libkernel\vec.hpp:897:17: note: declared private here
    897 |   VectorStorage _data;
        |                 ^
```

To address this, I modified the code as follows, explicitly constructing a new vector from the swizzled components:

`pos += sycl::float2(velocity3d.xy()) * delta_t / cell_size;`

This modification resolved the issue in AdaptiveCpp. However, when I compile the original code with oneAPI (DPC++), it compiles without any problems.

I'd like to know if this behavior in AdaptiveCpp is intended: probably there's something specific to SYCL standards that might explain this behavior?

Thank you for your time and assistance.

**Comment by illuhad** [comment]
Hi,

in short: I have no idea. Swizzles are very complicated and the specification is fairly vague with respect to how exactly they are implemented. This sort of issue typically comes up in the context of how exactly the constructor and implicit conversion operators are implemented. E.g. templates in the function signature can suppress implicit conversions.

Personal note: I would not use the `vec` class. It really brings barely any benefit but is well known for slowing down compile times substantially. SYCL compilers already vectorize horizontally across work items, so the use of `vec` is generally not needed for vectorization. Its use might lead to some vectorized loads and stores, but you can usually already achieve that easily yourself with your own structs with a suitable alignment.

**Comment by blinkfrog** [comment]
Thank you for sharing your thoughts and suggestions.

I vaguely recall reading (though I can't quite pinpoint whether it was in the context of DPC++, AdaptiveCpp, or SYCL in general) that swizzle operations may return something like a "swizzled_vector" and different SYCL implementations might handle implicit conversions to vectors differently.

Regarding the use of the `vec` class, I've thought about the possibility of using custom vector structs. However, there are some trade-offs to consider. If I were to switch to custom structs, I would lose the ability to seamlessly utilize built-in SYCL functions that operate directly on vectors, such as `length` or `distance`. There are also third-party solutions exist, for example, `glm`, which offers more user-friendly swizzles, a lot of useful functions, and a more compact syntax (comparing to SYCL) which is compatible with GLSL/HLSL. I've experimented with `glm` and found it to be quite capable, although it's not without its own challenges.

One issue I encountered with `glm` is that `glm::vec3` has an alignment of 12 bytes, in contrast to `sycl::float3`, which has an alignment of 16 bytes. This led to approximately a 50% slowdown in my tests in memory read/write access when using a buffer of vectors. This raises questions about the overall optimization of `glm` for parallel use, as well as its compatibility with other backends like CUDA or OpenCL, where native SYCL vectors and functions might map more optimally to their hardware equivalents.

Given these considerations, I've decided that using internal SYCL vectors and functions is probably a better idea than to use custom/third-party libraries.

Once again, thank you for your thoughts.

**Reply by illuhad** [reply]
> I vaguely recall reading (though I can't quite pinpoint whether it was in the context of DPC++, AdaptiveCpp, or SYCL in general) that swizzle operations may return something like a "swizzled_vector" and different SYCL implementations might handle implicit conversions to vectors differently.

That's right, the swizzle returns an implementation-defined "swizzled view" that should decay back to a regular `vec` when needed. How exactly that happens is pretty vague.

> Regarding the use of the vec class, I've thought about the possibility of using custom vector structs. However, there are some trade-offs to consider. If I were to switch to custom structs, I would lose the ability to seamlessly utilize built-in SYCL functions that operate directly on vectors, such as length or distance. There are also third-party solutions exist, for example, glm, which offers more user-friendly swizzles, a lot of useful functions, and a more compact syntax (comparing to SYCL) which is compatible with GLSL/HLSL. I've experimented with glm and found it to be quite capable, although it's not without its own challenges.

Right, I can see that `vec` can make the integration into SYCL more convenient. On the other hand, the built-in functions like `length` generally are not magic. You could probably build with similar performance using scalar builtins like `rsqrt`.

I haven't looked at `glm` so far. There arguments in favor and against 16-byte aligned `float3`. For example, AFAIR Gromacs have deliberately introduced their own `vec3` type because for them the waste of space was worse.

**Reply by blinkfrog** [reply]
Thank you for your suggestions. Well, the point was that I don't know exactly how SYCL code is mapped to OpenCL and CUDA, and also how CUDA code is mapped to hardware instructions, so I don't know if my (or third-party) custom code, mimicking built-in SYCL functions, especially functions which work with vectors, will be the same effective and will be mapped to the same hardware instructions as built-in SYCL functions.

(Also, as for `glm`, which always was my favorite library for scalar code. Seeing how developers make changes to the code to improve compatibility with SYCL and then revert them back because they introduced new problems doesn't add confidence - can I rely to it and use it in big long-term projects?)

**Reply by illuhad** [reply]
> Thank you for your suggestions. Well, the point was that I don't know exactly how SYCL code is mapped to OpenCL and CUDA, and also how CUDA code is mapped to hardware instructions, so I don't know if my (or third-party) custom code, mimicking built-in SYCL functions, especially functions which work with vectors, will be the same effective and will be mapped to the same hardware instructions as built-in SYCL functions.

Well, you can always try :) In general, the `vec` class and associated builtins are *not* magic. It should be possible to get the same perf with user code. Perhaps even better due to less C++ layers on top that might confuse the compiler at some point.
